libtenzir/builtins/functions/misc.cpp

```diff
@@ -228,6 +230,55 @@ class length final : public method_plugin {
   }
 };
 
+class has final : public method_plugin {
+public:
+  auto name() const -> std::string override {
+    return "has";
+  }
+
+  auto make_function(invocation inv, session ctx) const
+    -> failure_or<function_ptr> override {
+    auto expr = ast::expression{};
+    auto needle = located<std::string>{};
+    TRY(argument_parser2::method(name())
+          .add(expr, "<expr>")
+          .add(needle, "<string>")
+          .parse(inv, ctx));
+    return function_use::make([needle = std::move(needle),
+                               expr = std::move(expr)](evaluator eval,
+                                                       session ctx) -> series {
+      TENZIR_UNUSED(ctx);
+      auto value = eval(expr);
+      auto f = detail::overload {
+        [&](const arrow::NullArray& array) -> series {
+          auto b = bool_type::make_arrow_builder(arrow::default_memory_pool());
+          check(b->AppendValues(array.length(), false));
+          return {bool_type{}, finish(*b)};
+        },
+          [&](const arrow::StructArray& array) -> series {
+            const auto names = array.struct_type()->fields()
+                               | std::views::transform(&arrow::Field::name);
+            auto b
+              = bool_type::make_arrow_builder(arrow::default_memory_pool());
+            for (auto i = int64_t{0}; i < array.length(); i++) {
+              if (array.IsNull(i)) {
+                check(b->AppendNull());
+                continue;
+              }
```

Comment:
Should this append `false` or `null`?

Reply:
Probably `false` to be consistent with above.

Reply:
Good point, the behavior between `null` of `null_type` and `record_type` definitely needs to be the same. However, returning `null` in both cases would also be consistent.

I thought about this a bit more, and think that `null.has("foo") == null` makes more sense and is more aligned with the rest of our semantics. Would you agree?

Reply:
I think restricting the output to just booleans, makes it cleaner and easier to reason about. In other places, we usually don't have any valid data to return on a `null` input, but I think returning `false` makes sense here.

Reply:
I think we need to decide once whether `null.anything(...)` should ever be able to return anything but `null`. I think we can argue in both directions, but we should be consistent between functions.

Reply:
I think that generally we want to treat `null` the same way whether it comes from a `record` or a `null` type. So that leaves us with two options:
- If we return `null`, then we probably want to do a null check for records after all for each value, as otherwise `null.has(<str>)` would have a different return value depending on whether the null is a record type or not.
- If we return `false`, then we can omit the null check, and solely operate on the type itself.

Reply:
I think we have to check whether the method subject is null in any case. Otherwise, `foo.has("bar")` has different behavior when `foo` is `null_type` or `{bar: int64}`. In the table below, the italic values directly follow from the value to their left.

| &emsp;&emsp;&emsp;&emsp;foo = | null: null | null: {bar: int64} | {bar: 42}: {bar: int64} | 42: int64      |
|----------------|------------|--------------------|-------------------------|----------------|
| foo.has("bar") | null/false | _null/false_         | true                    | null (warning) |
| foo.has("baz") | null/false | _null/false_         | false                   | null (warning) |

Reply:
I think we should return `null` for consistency. Almost all other functions/methods do the same, and it would be weird to not do it here just because we have some semi-useful default value. I feel like the only exception should be functions/methods such as `foo.is_null()` (not really useful though, as it's the same as `foo == null`), or `foo.is_ip()` (which _could_ return `false` for `null`, but even there I'm not certain).